### PR TITLE
Template to send emails

### PR DESCRIPTION
**Pull request by phamkelly17**
# Notion ticket link

[Email Reminders](https://www.notion.so/uwblueprintexecs/Sistema-03b6ce1c4bc447b5945ab8b79cb00ea5?p=ae2ad9edf3b64073ae6dcc39a3fd3bc0&pm=s)

## Implementation description
- Use NodeMailer to be able to send emails
- Implemented unstyled template that allows us to send emails 

## Steps to test

1. Go to `http://localhost:3000/email`
2. Add recipients to the first text input. If you want to send to multiple emails, split them using a comma **with no space after the comma**. (e.g. "kellypham@uwblueprint.org, jasonling@uwblueprint.org")
3. Add the email body to the textbox
4. Click send 
5. Check your email and you should have received an email from kellypham@uwblueprint.org

## What should reviewers focus on?

- functionality

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -53,4 +54,4 @@
       "git add schema.prisma"
     ]
   }
-}
\ No newline at end of file
+}
```
Comment (ChinemeremChigbo) [inline package.json:57]:
```suggestion
}
```

File: src/pages/api/sendEmail.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable prettier/prettier */
```
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:1]: Why disable prettier? It's generally better to allow Prettier to format the code for consistency and readability.

File: src/pages/email.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// * eslint-disable prettier/prettier */
```
Comment (ChinemeremChigbo) [inline src/pages/email.tsx:1]: Same prettier comment.

File: src/pages/index.tsx [hunk 4]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable prettier/prettier */
```
Comment (ChinemeremChigbo) [inline src/pages/index.tsx:1]: Same prettier comment.

File: package.json (hunk 1 above)
Comment (ChinemeremChigbo) [inline package.json:57]: I would not add the extra blank line.

File: src/pages/api/sendEmail.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import { NextApiRequest, NextApiResponse } from 'next';
+
+const nodemailer = require('nodemailer');
```
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:4]: Mixing require and import statements can lead to inconsistencies. I would use consistent syntax. i.e `import nodemailer from 'nodemailer';`

File: src/pages/api/sendEmail.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  try {
+    const transporter = nodemailer.createTransport({
+      service: 'gmail',
+      auth: {
+        user: ['kellypham@uwblueprint.org'],
+        pass: 'spnu pjuc zvxx subj', // app password
+      },
+    });
+
+    // Configure the mailoptions object
```
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:28]:
```suggestion
    // Configure the mailOptions object
```

File: src/pages/api/sendEmail.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+};
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<Data>
+) {
+  const { to, subject, text } = req.body;
+
+  // Create a transporter object
+  try {
```
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:19]: I would first do some basic input validation.

File: src/pages/api/sendEmail.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    const transporter = nodemailer.createTransport({
+      service: 'gmail',
+      auth: {
+        user: ['kellypham@uwblueprint.org'],
```
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:23]: I would put these variables in the .env file instead of hardcoding them.

File: src/pages/api/sendEmail.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        pass: 'spnu pjuc zvxx subj', // app password
+      },
+    });
+
+    // Configure the mailoptions object
+    const mailOptions = {
+      from: 'kellypham@uwblueprint.org',
```
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:30]: I would put this in the .env file instead of hardcoding it

File: src/pages/api/sendEmail.ts (hunk 8 above)
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:23]: I'm not sure why this is a list when I think use should be a string.

File: src/pages/api/sendEmail.ts
Comment (ChinemeremChigbo) [inline src/pages/api/sendEmail.ts:1]: From my feedback, I think something like this would be better feel free to push back on any points you disagree with!

``` ts
import { NextApiRequest, NextApiResponse } from 'next';
import nodemailer from 'nodemailer';

type Data = {
  message?: string;
  error?: string;
  details?: string;
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {
  const { to, subject, text } = req.body;

  // Validate the input
  if (!to || !subject || !text) {
    return res.status(400).json({ error: 'Missing required fields' });
  }

  try {
    // Create a transporter object using environment variables
    const transporter = nodemailer.createTransport({
      service: 'gmail',
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASS,
      },
    });

    // Configure the mailOptions object
    const mailOptions = {
      from: process.env.EMAIL_USER,
      to,
      subject,
      text,
    };

    // Send the email
    await transporter.sendMail(mailOptions);

    res.status(200).json({ message: 'Email sent successfully' });
  } catch (error: any) {
    res
      .status(500)
      .json({ error: 'Error sending email', details: error.message });
  }
}
```

File: src/pages/email.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import { useState } from 'react';
+import { Button, Container, Input, Text, Textarea } from '@chakra-ui/react';
+
+const Email = () => {
+  const [loading, setLoading] = useState(false);
+  const [message, setMessage] = useState<string>('');
+  let [emailText, setEmailText] = useState('');
```
Comment (ChinemeremChigbo) [inline src/pages/email.tsx:8]: Should be const

File: src/pages/email.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  let [emailRecipients, setEmailRecipients] = useState([]);
```
Comment (ChinemeremChigbo) [inline src/pages/email.tsx:9]: Should be const

File: src/pages/email.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    let recipientsString = e.target.value;
+    setEmailRecipients(recipientsString.split(','));
+  };
+
+  return (
+    <Container>
+      <Input
+        placeholder="Add recipients here"
+        onChange={handleEmailRecipientsChange}
+      ></Input>
```
Comment (ChinemeremChigbo) [inline src/pages/email.tsx:55]: I usually use self-closing tags for elements without children

File: src/pages/email.tsx [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+      <Textarea
+        placeholder="Write your email here"
+        value={emailText}
+        onChange={handleEmailTextChange}
+      ></Textarea>
```
Comment (ChinemeremChigbo) [inline src/pages/email.tsx:60]: I usually use self-closing tags for elements without children

File: src/pages/email.tsx
Comment (ChinemeremChigbo) [inline src/pages/email.tsx:1]: Here is an example of the file with most of the changes I mentioned I applied. Feel free to push back on what you disagree with:

```
import { useState } from 'react';
import { Button, Container, Input, Text, Textarea } from '@chakra-ui/react';

const Email = () => {
  const [loading, setLoading] = useState(false);
  const [message, setMessage] = useState<string>('');
  const [emailText, setEmailText] = useState('');
  const [emailRecipients, setEmailRecipients] = useState<string[]>([]);

  const handleEmailSend = async () => {
    setLoading(true);
    setMessage('');
    try {
      const response = await fetch('/api/sendEmail', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          to: emailRecipients,
          subject: 'Hello World',
          text: emailText,
        }),
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.details || 'Failed to send email');
      }

      const data = await response.json();
      setMessage(data.message);
    } catch (error: any) {
      setMessage(`Error sending email: ${error.message}`);
    } finally {
      setLoading(false);
    }
  };

  const handleEmailTextChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setEmailText(e.target.value);
  };

  const handleEmailRecipientsChange = (
    e: React.ChangeEvent<HTMLInputElement>
  ) => {
    const recipientsString = e.target.value;
    setEmailRecipients(recipientsString.split(','));
  };

  return (
    <Container>
      <Input
        placeholder="Add recipients here"
        onChange={handleEmailRecipientsChange}
      />
      <Textarea
        placeholder="Write your email here"
        value={emailText}
        onChange={handleEmailTextChange}
      />
      <Button
        onClick={handleEmailSend}
        isLoading={loading}
        loadingText="Sending"
        colorScheme="blue"
      >
        Send Email
      </Button>
      {message && (
        <Text
          mt={4}
          color={message.startsWith('Error') ? 'red.500' : 'green.500'}
        >
          {message}
        </Text>
      )}
    </Container>
  );
};

export default Email;
```

File: src/pages/api/sendEmail.ts (hunk 8 above)
Comment (ChinemeremChigbo) [reply src/pages/api/sendEmail.ts:23]: Please contact me to put the email / password in the hashicorp vault. I feel like we should probably use the sistema email instead of kelly's so let me know when you would like to set that up!

File: tsconfig.json [hunk 14]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "target": "es5",
+    "target": "es2017",
```
Comment (ChinemeremChigbo) [inline tsconfig.json:3]: Maybe we should change the ecmaScript version, but I feel like this should be in a different PR.

File: src/pages/api/sendEmail.ts (hunk 2 above)
Comment (phamkelly17) [reply src/pages/api/sendEmail.ts:1]: whoopsies copied that in from the index page, it was prettier disabled

File: src/pages/api/sendEmail.ts (hunk 8 above)
Comment (phamkelly17) [reply src/pages/api/sendEmail.ts]: done

**Comment by ChinemeremChigbo** [comment]
Please add the following to the .env.sample file:

EMAIL_PASS=
EMAIL_SERVICE=
EMAIL_USER=

File: tsconfig.json (hunk 14 above)
Comment (ChinemeremChigbo) [inline tsconfig.json:3]: I would just change this back to es5. Other than that, LGTM.

**Review by ChinemeremChigbo — APPROVED** [review]

File: src/pages/api/sendEmail.ts [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+  details?: string;
+};
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<Data>
+) {
+  const { to, subject, text } = req.body;
+
+  // Validate the input
```
Comment (RahulTandon1) [inline src/pages/api/sendEmail.ts:16]: Do you it makes sense to use a validation library like [Zod](https://zod.dev/) or [Joi](https://joi.dev/api/?v=17.13.3) instead of writing this ourselves?

File: src/pages/email.tsx [hunk 16]
```diff
@@ -0,0 +1,80 @@
+import { useState } from 'react';
+import { Button, Container, Input, Text, Textarea } from '@chakra-ui/react';
+
+const Email = () => {
+  const [loading, setLoading] = useState(false);
+  const [message, setMessage] = useState<string>('');
+  const [emailText, setEmailText] = useState('');
+  const [emailRecipients, setEmailRecipients] = useState([]);
+
```
Comment (RahulTandon1) [inline src/pages/email.tsx:9]: I'm not sure what's the best way of integrating TypeScript with useState.
I do believe that being consistent with usage is important. If we want to use `useState<type>`, could we please add them to all the states here?

File: src/pages/email.tsx [hunk 17]
```diff
@@ -0,0 +1,80 @@
+import { useState } from 'react';
+import { Button, Container, Input, Text, Textarea } from '@chakra-ui/react';
+
+const Email = () => {
+  const [loading, setLoading] = useState(false);
```
Comment (RahulTandon1) [inline src/pages/email.tsx:5]: nit: could we please rename this to `isLoading` since it's a boolean?

File: src/pages/email.tsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    setMessage('');
+    try {
+      const response = await fetch('/api/sendEmail', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          to: emailRecipients,
+          subject: 'Hello World',
```
Comment (RahulTandon1) [inline src/pages/email.tsx:21]: Could we add another Input/TextArea in the form and get the subject from there?

File: src/pages/api/sendEmail.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+        pass: process.env.EMAIL_PASS, // app password
+      },
+    });
+
+    // Configure the mailOptions object
+    const mailOptions = {
+      from: process.env.EMAIL_USER,
+      to: to,
+      subject: subject,
+      text: text,
```
Comment (RahulTandon1) [inline src/pages/api/sendEmail.ts:36]: It may be useful to also add `html: text` here.
According [this page](https://nodemailer.com/message/) of the Nodemailer docs, it also supports HTML. I think eventually, we would want to enable the admin to style things as well.
@ChinemeremChigbo could you please confirm that this is within the scope? if not, this comment could be resolved directly.

File: src/pages/email.tsx [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+  const handleEmailRecipientsChange = (e) => {
+    const recipientsString = e.target.value;
+    setEmailRecipients(recipientsString.split(','));
+  };
+
+  return (
+    <Container>
+      <Input
+        placeholder="Add recipients here"
```
Comment (RahulTandon1) [inline src/pages/email.tsx:52]: Nothing to address here. For my knowledge's sake: I would usually expect:
```react
<Input placeholder="random placeholder" value={value}>
```
but we were able to do without this. How does this happen?
Does React not always need this `value={stateVar}` thing?
Would appreciate any feedback/links/something to read up

**Comment by RahulTandon1** [comment]
@ChinemeremChigbo I needed some clarification about the ticket.

**Review by RahulTandon1 — CHANGES_REQUESTED** [review]
It's really cool that you were able to get this running with such succinct code. I was expecting NodeMailer to be really tough to get working, but you made it look like easy like a pro dev :-)

File: src/pages/email.tsx (hunk 16 above)
Comment (phamkelly17) [reply src/pages/email.tsx:9]: done, removed useState<string>('');

File: src/pages/email.tsx (hunk 18 above)
Comment (phamkelly17) [reply src/pages/email.tsx:21]: done

File: src/pages/api/sendEmail.ts (hunk 19 above)
Comment (phamkelly17) [reply src/pages/api/sendEmail.ts:36]: I think the scope of this ticket was just to set up nodemailer and send an email, and we'll deal with email format later because as of right now there's no specification of what the email should look like

File: src/pages/api/sendEmail.ts (hunk 15 above)
Comment (phamkelly17) [reply src/pages/api/sendEmail.ts:16]: From my understand, these emails will sent based off the email in the User object. So the validation should happen there so that there are no invalid emails in our database. So I don't think any validation needs to happen here. Lmk if you think otherwise.

**Comment by phamkelly17** [comment]
@RahulTandon1 to answer your questions:
Q1: I thought the email would be a generic message like "Hi `name`, we see you haven't uploaded your lesson plan for your lesson in `num` days. Sincerely, Sistema". So I didn't think the template would be configurable.
Q2: Correct, this ticket is solely for functionality

File: src/pages/email.tsx (hunk 20 above)
Comment (phamkelly17) [reply src/pages/email.tsx:57]: I think for the Input element in HTML it just knows to store whatever you type in

File: src/pages/api/sendEmail.ts (hunk 15 above)
Comment (RahulTandon1) [reply src/pages/api/sendEmail.ts:16]: Apologies for replying late.
- imo ideally some basic validation would be a good idea as a sanity check. If we ever write incorrect code such that the DB isn't populated correctly, being able to detect it here via a log message sounds useful to me. But this just an opinion, not a fact.

- Could you please tell me whether the following is correct?
Eventually, we'll be sending emails from a server side piece of code only. So, the current handler function won't be called by the client-side code directly.

File: tsconfig.json (hunk 14 above)
Comment (ChinemeremChigbo) [reply tsconfig.json]: Could you change it back lol.

**Comment by ChinemeremChigbo** [comment]
> @ChinemeremChigbo could you please help me understand the ticket a bit more.
> 
> To the best of my understanding, we're trying to do the following: Whenever an absence is created/claimed (or edited in any important way), we want to notify all _relevant_ users with an email.
> 
> The email follows a template. Example :
> 
> ```
> // not this does not follow the DB schemas. I'm just made the fields up.
> 
> Hey {user.firstName}!
> The absence for the class {absence.class}  on `absence.date` has been claimed by {user2.firstName} {user2.lastName}.
> More text...
> 
> The Sistema Toronto Team
> ```
> 
> Q1: In the long term / actual product we want the admin to be able to modify this template through a GUI...right? Q2: The main point of this ticket is to get the email sending functionality up and running...right? Q3: if there's anything you think is wrong with my understanding, could you please correct it?

1. Yes
2. Yes
3. There's nothing to correct

File: src/pages/api/sendEmail.ts (hunk 15 above)
Comment (RahulTandon1) [reply src/pages/api/sendEmail.ts:16]: Talked to Chinemerem. My bad. Don't need to deal with this rn. Will deal with this issue when we're integrating this code. Thank you Kelly!

**Review by RahulTandon1 — APPROVED** [review]
We're good to ship. 
Sorry for all my delays Kelly.
Thank you!
